### DISCUSSION TITLE
FAQ

### DISCUSSION THREAD

**Discussion by xnorpx** [xnorpx/blue-onyx#14] [Documentation]
---

**Q: Will Blue Onyx ever support TPUs (Coral TPUs)?**  
**A:** Blue Onyx can support any backend that the ONNX Runtime supports. Currently, ONNX does not support Coral TPUs: [[ONNX Runtime Issue #10248](https://github.com/microsoft/onnxruntime/issues/10248)]. So please go bring up Coral support in the issue above.

---

**Q: How can I run multiple object detection models at the same time in one instance of Blue Onyx**
**A:** Blue Onyx is designed to run one detection model. If you want to run multiple Blue Onyx instances with different models on different ports that possible. The reason is the complexity it will bring into the fairly simple design, and I think it is not worth the effort for most use cases.

---

**Q: My Datacenter NVidia card doesn't work!**
**A:** Datacenter NVidia card doesn't support DirectX since they are datacenter cards. However you can play around with driver versions and register edits see: https://github.com/xnorpx/blue-onyx/issues/90

---

**Q: Windows Defender says it is a virus!**  
**A:** Some users have reported that Windows Defender flags it as a virus. Please verify this by checking the file on [[VirusTotal](https://www.virustotal.com/)](https://www.virustotal.com/) to ensure it’s a false positive.

---

**Q: Error: An attempt was made to access a socket in a way forbidden by its access permissions. (OS error 10013)**  
**A:** It appears that another application is already using that port, such as another instance of CPAI or Blue Onyx. Try using a different port, for example, by adding `--port 32167`.

---

**Q: I don't have a GPU so it fails with ERROR ort::execution_providers: An error occurred when attempting to register DmlExecutionProvider: blue-onyx\blue-onyx\target\release\build\blue-onyx-2235df632c69ed17\out\onnxruntime-1.20.1\onnxruntime\core\providers\dml\dml_provider_factory.cc(491)\onnxruntime.dll!00007FFBC1F8B973: (caller: 00007FFBC1F8BC36) Exception(1) tid(2d70) C0262002 Specified display adapter handle is invalid.**
**A.** Use cli --force-cpu and it will not try to run on GPU.

---

**Q: I have multiple GPUs in my system and the GPU-Index does not list as correct device when I pick it?**  
**A:** There is no useful feedback on which GPU has been selected based on the index. In case of a mismatch, you have to determine for yourself which GPU index corresponds to which GPU.

---

**Q: Why does the inference time seem much worse when I run it as a service compared to the benchmarks?**  
**A:** Inference tends to perform better under continuous load, which allows it to run more efficiently than handling ad hoc requests.

---

**Q: Do you earn money from this project?**  
**A:** No, I use this personally as an alternative to CPAI and decided to share it as open source. I may consider adding sponsorship options later if the project proves useful to others, to help cover costs like coffee money.

---

**Q: Does Blue Onyx offer all the features that CodeProject.AI (CPAI) has: [[CodeProject.AI Server - AI the Easy Way](https://www.codeproject.com/Articles/5322557/CodeProject-AI-Server-AI-the-easy-way)](https://www.codeproject.com/Articles/5322557/CodeProject-AI-Server-AI-the-easy-way)?**  
**A:** Not exactly. Blue Onyx is designed to run a single DETR model on either CPU or GPU as simply and efficiently as possible.

---

**Q: Why not just use CPAI?**  
**A:** I spent countless hours installing CPAI, dealing with broken updates, and managing thousands of Python files. I became frustrated with the complexity and wanted to create something simpler and more stable to solve my problem: image input, bounding box output without the hassle.

---

**Q: Will you add more features to make it more like CPAI?**  
**A:** Probably not. My goal with Blue Onyx is to stay focused on doing one thing exceptionally well without adding unnecessary features or increasing maintenance. However, if many people request a specific feature, it can be discussed.

---

**Q: There have been no releases for three months! Is the project dead?**  
**A:** No. If the project were dead, it would be marked as deprecated on GitHub. New releases will occur when there are new features, bug fixes, or security updates. Critical dependency updates can also trigger a new release. I have a full-time job and family commitments, so time for side projects is limited. However, anyone is welcome to fork the code and make updates.

---

**Q: I’d like to contribute!**  
**A:** That’s great! For major features, please open a discussion. For smaller bug fixes, feel free to submit a pull request.

---

**Q: Why Rust?**  
**A:** Rust is enjoyable to work with and generally ensures very stable code.

---

**Q: What inference backend is Blue Onyx using?**  
**A:** It uses the [[ONNX Runtime](https://github.com/microsoft/onnxruntime)](https://github.com/microsoft/onnxruntime) from Microsoft.

---

**Q: Why ONNX?**  
**A:** ONNX is an industry standard with an actively maintained team and tools available for converting PyTorch models to ONNX format.

---

**Q: But ONNX is written in C++, not Rust!**  
**A:** Yes, but C++ can also be safe, secure, and performant. It’s just more challenging, and there is a dedicated team at Microsoft handling that work.

---

**Q: Wait a minute, Microsoft says they might collect telemetry data!**  
**A:** Blue Onyx builds ONNX from source, which does not include telemetry data collection, so you’re all set.

---

**Q: What ONNX backends is Blue Onyx built with?**  
**A:** Currently, it uses [[DirectML](https://github.com/microsoft/directml)](https://github.com/microsoft/directml), but it can also run on CPU.

---

**Q: What about CUDA, TensorRT, and OpenVINO backends?**  
**A:** DirectML supports most GPUs in a generic way, making it easy to build and integrate while supporting multiple GPU vendors on Windows. Pull requests to support more backends are welcome!

---

**Q: My X-year-old Intel GPU is slow and performs worse than I expected!**  
**A:** DETR models are large and complex to run. While DirectML allows the model to run on a variety of hardware, it won’t magically make older integrated GPUs fast.

---

**Q: I have a Linux server. Does Blue Onyx work on Linux?**  
**A:** Currently, it only builds for Windows. Porting it to Linux should be feasible, and pull requests are welcome.

---

**Q: Can Blue Onyx use custom-trained models?**  
**A:** Yes, Blue Onyx supports using custom-trained DETR models in ONNX format.

---

**Q: Will you provide custom models for <insert an object you want to detect here>?**  
**A:** No, training custom models is outside the scope of Blue Onyx, but it should support running them.

---

**Q: Why DETR models and not YOLO 8+?**  
**A:** Firstly, YOLO has an AGPL license from some sources, which creates a gray area for open source projects. I wanted to move away from that. Secondly, I aim to run state-of-the-art models, and DETR seems to be the future in this area. However Blue Onyx does support Yolo5 models.

For more discussion on this subject, see the following link: [[YouTube Discussion](https://www.youtube.com/live/wT636THdZZo?si=00syQ5xAVTgMhUJl&t=5619)](https://www.youtube.com/live/wT636THdZZo?si=00syQ5xAVTgMhUJl&t=5619)

---

**Comment by BillHeacock** [comment]
I've been waiting for some of the dust to settle with Blue Iris and video recognition, and I'm a patient man. 
It has been moving forward nicely. 
Reading about Blue Onyx has resonated with me finally. I'll be diving in shortly. 
Thanks for your efforts and the open source stance. I think you're onto something users can build upon!
